include/envoy/http/header_map.h

```diff
@@ -310,6 +310,7 @@ class HeaderEntry {
   HEADER_FUNC(Etag)                                                                                \
   HEADER_FUNC(Expect)                                                                              \
   HEADER_FUNC(ForwardedClientCert)                                                                 \
+  HEADER_FUNC(ForwardedUntrustedClientCert)                                                        \
```

Comment:
Do we really want a new header field for untrusted client certificates? Can we just add `Trusted=False` in XFCC header to indicate that?

Reply:
We had initially implemented this way, however had concerns over the potential for existing code to treat ForwardedClientCert as having passed validation.
An example of the routing we perform is as follows: - route to the intended service, iff have x-forwarded-client-cert header. (I guess a 'safe-regex_match' could alternatively be used to route based on 'Trusted=False/True'

"routes": [
    {
      "match": { "prefix": "/", "headers": [
        { "name": "x-forwarded-untrusted-client-cert", "present_match": true }
       ] },
      "route": { 
        "cluster": "venue-edge-access-control",
        "prefix_rewrite": "/api/cert/untrusted",
        "host_rewrite": "venue-edge-access-control"
       }
    },
    {
      "match": { "prefix": "/", "headers": [
        { "name": "x-forwarded-client-cert", "present_match": true, "invert_match" : true }
       ] },
       "route": { 
        "cluster": "venue-edge-access-control",
        "prefix_rewrite": "/api/cert/untrusted",
        "host_rewrite": "venue-edge-access-control"
       }
    },
    {
      "match": { "prefix": "/", "headers": [
        { "name": "x-forwarded-client-cert", "present_match": true },
        { "name": ":authority", "prefix_match": "service-actual-target-service" }
       ] },
      "route": {
        "cluster": "service-actual-target-service"
      }
    },
    ....

Reply:
@PiotrSikora thoughts on these headers and what other proxies might do.

Reply:
@jimini-lumox I see what your example try to address. Does it make more sense to have a route matcher based on presented client cert semantically rather than relying on XFCC/XFUCC header? Adding header and route matching can be orthogonal.

Reply:
I'm not sure I understand exactly, or how to achieve it using the route matcher. Below is a comment on our usage from the first cut of the PR (https://github.com/envoyproxy/envoy/pull/5207). Do you know how we would use route matching to handle the scenarios (maybe it's easy, I'm not sure)

We use Envoy as an Edge Proxy to host services, with thousands of devices in venues on a private network.
We're using the Hash List as validation of clients, which will initially be empty ie - nothing to validate against.
Client Certs are currently self-signed (no PKI infrastructure available) - so there is nothing within the certificate to validate against.
If a client connects and its Cert Hash is in the list, it will be routed to its intended target service.
(eg: x-forwarded-client-cert header present & :authority header match)
If a client connects and its Cert Hash is not in the list (or there is no validation context yet), then it may be routed to an access-control-service.
(eg: x-forwarded-untrusted-client-cert header present & :authority header match)
Then the access-control-service controls whether a client is to be permitted, in which case the client hash will be added to the verify_certificate_hash list (using dynamic_resources from the Edge Proxy, so subsequent request is routed to the intended service).
Additionally we may not care if a client needs to be validated to reach certain other host services.

Reply:
We can change it to use the x-forwarded-client-cert with 'Trusted=False' and match on this.

Reply:
I mean, add a field in [RouteMatch](https://github.com/envoyproxy/envoy/blob/master/api/envoy/api/v2/route/route.proto#L324) to do matching based on the client certificate, we already do similar thing in [RBAC](https://github.com/envoyproxy/envoy/blob/master/api/envoy/config/rbac/v2/rbac.proto#L171), so adding PrincipalMatch (name TBD) in RouteMatch to do first class matching of certificate will be cleaner IMO. WDYT?

Reply:
RouteMatch only uses http headers, so if I'm following you correctly:
* always hoist headers into x-forwarded-client-cert (instead of using x-forwarded-untrusted-client-cert for either failed or unvalidated)
* and,
  either add a field to the x-forwarded-client-cert to indicated that it passed/failed validation (eg Trusted=false)
  or add another http header that simply indicates whether the client was validated, so we can route based on this.
* and then add a new RouteMatch field that can simplify the route matching rules for validation, eg something like:

  message RouteMatch {
    message CredentialMatchOptions {
        google.protobuf.BoolValue validated = 1;
        // can add further matches here - eg Issuer,Hash etc.
    }
    ...
    CredentialMatchOptions credentials = 11;
  }
    
then the matching would simplify to:
```
"routes": [
  {
    "match": { "prefix": "/", "credentials": { "validated": false } },
    "route": {
      "cluster": "venue-edge-access-control",
      ...
    }
  },
  {
    "match": { "prefix": "/", "credentials": { "validated": true },
               "headers": [ { "name": ":authority", "prefix_match": "service-actual-target-service" }
               ]
    },
    "route": {
      "cluster": "service-actual-target-service"
    }
  }
...
```

Reply:
The config example looks right, I think we should open another issue for this API proposal.

> RouteMatch only uses http headers

This is how current implementation does but not a hard limitation. You can use the cert information directly from downstream connection and plumb it into RouteMatcher.

Reply:
Ok, so should I create a new PR that has this API proposal including the changes to route the required connection info through to the Route matching.
I was thinking that we could provide the StreamInfo through the route matching from which we can get the downstreamSslConnection info etc.

Reply:
> I was thinking that we could provide the StreamInfo through the route matching from which we can get the downstreamSslConnection info etc.

Exactly.

Reply:
I've raised PR #8453 with an initial cut of passing the connection stream_info through to the route matching.